src/vsr/routing.zig

```diff
@@ -0,0 +1,877 @@
+//! 5. Topology doesn't change;
+//! 6. There is one administrator;
+//! 7. Transport cost is zero;
+//! 8. The network is homogeneous;
+//!
+//! Robust tail principle:
+//! - The same code should handle slow nodes and crashed nodes.
+//! - A single crashed node should not cause retries and risk metastability.
+//!
+//! Algorithm:
+//!
+//! The replication route is V-shaped. Primary is in the middle, and it tosses each prepare at two
+//! of its neighbors. Backups forward prepares to at most one further backup. Neighbors of the
+//! primary have at least one more neighbor (in a six-replica cluster). If any single node fails,
+//! the primary still gets a replication quorum. If the primary and backups disagree about the
+//! replication route, the primary still gets a replication quorum.
+//!
+//! Because topology changes, routes are dynamic. The primary broadcasts the current route in the
+//! ping message. It's enough if the routing information is only eventually consistent.
+//!
+//! To select the best route, primary uses outcome-focused explore-exploit approach. Every once in a
+//! while the primary tries an alternative route. The primary captures replication latency for a
+//! route (that is, the arrival time of prepare_ok messages). If the latency for an alternative
+//! route is sufficiently better than current latency, the route is switched. Note that latency
+//! includes both network and disk latency.
+//!
+//! The experiment schedule is defined randomly. All replicas share the same RNG seed, so no
+//! coordination is needed to launch an experiment!
+//!
+//! To remove outliers, the experiment take two ops. op and op|1 are either both experimental, or
+//! both non-experimental. Experiment's cost is the average of two costs. The cost for a route
+//! is exponential moving average.
+const std = @import("std");
+const assert = std.debug.assert;
+const maybe = stdx.maybe;
+const constants = @import("../constants.zig");
+const stdx = @import("../stdx.zig");
+const Ratio = stdx.PRNG.Ratio;
+const ratio = stdx.PRNG.ratio;
+const Instant = stdx.Instant;
+const Duration = stdx.Duration;
+
+const history_max = constants.pipeline_prepare_queue_max * 2;
+
+// This constant serves two purposes:
+// - First, as a cap for all latencies, to make sure that u64 computations can't overflow.
+// - Second, to simplify cost function computations, replica we haven't heard from has this latency.
+const latency_max: Duration = .{ .ns = 10 * std.time.ns_per_hour };
+
+replica: u8,
+replica_count: u8,
+standby_count: u8,
+view: u32,
+
+/// The route we currently use.
+a: Route,
+a_cost: ?Cost, // Computed as exponential weighted moving average.
+
+/// The best alternative, which might be better or worse than a.
+b: Route,
+b_cost: ?Cost, // Computed as an average of two experiments.
+
+experiment_chance: Ratio = ratio(1, 20),
+
+history: [history_max]OpHistory,
+
+pub const Route = stdx.BoundedArrayType(u8, constants.replicas_max);
```

Comment:
I find it more readable to declare fields before types, as fields are more important.